### PR TITLE
Use`include` with the correct order via `@require` in `__init__`

### PR DESCRIPTION
**Pull request by devmotion**
Fixes https://github.com/TuringLang/AdvancedHMC.jl/issues/203.

### CODE REVIEW

**Comment by xukai92** [comment]
This PR still uses `include` in `__init__` right?

**Comment by devmotion** [comment]
Yeah but in the correct order. If the `include` statement is inside of the `@require` block, then Requires takes care of everything.

**Comment by devmotion** [comment]
See https://github.com/JuliaPackaging/Requires.jl#requiresjl

**Comment by xukai92** [comment]
Got it. I got confused because of the title. I changed it to something more clear. I will merge this if Travis is happy.

**Comment by devmotion** [comment]
:+1: Yes, indeed, the title was a bit misleading. Tests passed locally on Julia 1.0, so hopefully CI tests pass now :crossed_fingers:

**Comment by devmotion** [comment]
It looks quite good (most test failures are gone) but it seems one test fails since the samples are a bit off:
https://travis-ci.com/github/TuringLang/AdvancedHMC.jl/jobs/342167680#L675

**Comment by devmotion** [comment]
Oh I just noticed that the test failure that was left is also present on master.

**Comment by devmotion** [comment]
Now some other stochastic error showed up, even though the seeds are already fixed there.

File: src/ad.jl [hunk 1]
```diff
@@ -0,0 +1,18 @@
+const ADSUPPORT = (:ForwardDiff, :Zygote)
```
Comment (yebai) [inline src/ad.jl:1]: Please consider keeping this file under `contrib` folder, i.e. `contrib/ad.jl`. This is mostly to keep the ad and other misc stuff clearly separated from HMC code.  

The folder `contrib` is a bit abused at the moment, we might want to add a `utilities` folder in the future.

**Comment by devmotion** [comment]
https://github.com/devmotion/AdvancedHMC.jl/blob/73be302ffb3fac38900febacc32c34921eed10c0/test/sampler.jl#L88 still errors randomly in one run (Windows x64, Julia nightly) even though a seed is set. I guess that's due to the fact that RNGs are not guaranteed to produce the same numbers in different Julia versions and OS, even if they are initialized with the same seed.

File: src/ad.jl (hunk 1 above)
Comment (devmotion) [reply src/ad.jl]: OK, I changed it :+1:

**Comment by devmotion** [comment]
Tests still fail for StanHMCAdaptor, even with 20000 samples. Is there anything else that could be tuned?

**Comment by yebai** [comment]
I think the default max warmup iteration is 1000, so increasing the total sampling iterations won’t help. You might have to manually increase the number of warmup iterations. 

Also, I’m happy to merge because this is a failure on Julia’s nightly build. We can come back to this if there is a failure on stable releases.

File: test/sampler.jl [hunk 2]
```diff
@@ -9,7 +9,7 @@ include("common.jl")
 θ_init = rand(MersenneTwister(1), D)
 ϵ = 0.2
 n_steps = 10
-n_samples = 12_000
+n_samples = 22_000
 n_adapts = 2_000
```
Comment (yebai) [inline test/sampler.jl:13]:
```suggestion
n_adapts = 4_000
```

File: test/sampler.jl [hunk 3]
```diff
@@ -85,7 +85,7 @@ end
                     # For other adapatation methods that are able to adpat the step size, we use `find_good_stepsize`.
                     τ_used = adaptorsym == :MassMatrixAdaptorOnly ? τ : reconstruct(τ, integrator=reconstruct(lf, ϵ=find_good_stepsize(h, θ_init)))
                     samples, stats = sample(h, τ_used , θ_init, n_samples, adaptor, n_adapts; verbose=false, progress=PROGRESS)
-                    @test mean(samples) ≈ zeros(D) atol=RNDATOL
+                    @test mean(samples[(n_adapts+1):end]) ≈ zeros(D) atol=RNDATOL
```
Comment (xukai92) [inline test/sampler.jl:88]: Maybe use `drop_warmup=true` in the `sample` call above.

File: test/sampler-vec.jl [hunk 4]
```diff
@@ -51,9 +54,12 @@ include("common.jl")
         ]
             τ isa HMCDA && continue
             @test show(adaptor) == nothing
+
+            Random.seed!(100)
             samples, stats = sample(h, τ, θ_init_list[i_test], n_samples, adaptor, n_adapts; verbose=false, progress=false)
```
Comment (xukai92) [inline test/sampler-vec.jl:59]: Maybe `drop_warmup=true` here too.

**Comment by yebai** [comment]
I played with stepsize a bit, but don't have much luck. A quick inspection of the failed assertion suggest the error bound might be too tight: 

```
julia> [-0.142056, -0.624472, 0.0118594, -0.450631, 0.328566] ≈ [0.0, 0.0, 0.0, 0.0, 0.0]
false

julia> a = [-0.142056, -0.624472, 0.0118594, -0.450631, 0.328566]
5-element Array{Float64,1}:
 -0.142056 
 -0.624472 
  0.0118594
 -0.450631 
  0.328566 

julia> using LinearAlgebra

julia> norm(a)
0.8492995618775274
```

File: test/sampler.jl (hunk 3 above)
Comment (yebai) [reply test/sampler.jl:88]: yes, but the effect should be similar to what we currently do.

**Pull request merged**